analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/MetadataUtils.java

```diff
@@ -0,0 +1,273 @@
+
+        Long fileId = getFileId(pckVersionId, path);
+        List<Long> modulesId = getModuleIds(fileId);
+
+        ArrayList<CallableHolder> callables = new ArrayList<CallableHolder>();
+
+        if(!modulesId.isEmpty()) {
+
+            for(Long moduleId : modulesId) {
+                callables.addAll(getCallablesInformation(moduleId, lineStart, lineEnd));
+            }
+
+            //TODO: we need a unique record here!
+            logger.info("Found " + callables.size() + " methods for which startLine= " + lineStart + " and endLine= " + lineEnd);
+
+        }
+
+        return callables;
+
+    }
+
+    /**
+     * Retrieves the package_version_id given the purl of the package version.
+     * @param purl - follows purl specifications
+     * @return negative if it cannot be found
+     */
+    private Long getPackageVersionId(String coordinate, String forge, String version) {
+
+        logger.info("Looking for package_version_id of " + coordinate);
+
+        Long packageId = getPackageIdFromCoordinate(
+                coordinate,
+                forge);
+
+        if (packageId != null) {
+            return getPackageVersionIdFromVersion(
+                    packageId,
+                    version);
+        }
+
+        return -1L;
+    }
+
+
+
+    /**
+     * Finds the ID of the package given coordinate and forge.
+     * This is ecosystem agnostic
+     * @param coordinate - includes information about the package
+     * @param forge - ['mvn', 'PyPI', 'Debian']
+     * @return - Value of the package ID if found, null otherwise.
+     */
+    private Long getPackageIdFromCoordinate(String coordinate, String forge) {
+
+        PackagesRecord record = (PackagesRecord) selectedContext.select()
+                .from(Packages.PACKAGES)
+                .where(Packages.PACKAGES.PACKAGE_NAME.equal(coordinate))
+                .and(Packages.PACKAGES.FORGE.equal(forge))
+                .fetchOne();
+
+        if (record != null) {
+            return record.component1();
+        }
+
+        return null;
+    }
+
+    /**
+     * Finds the ID of the package_version in all the version of the package.
+     * @param pkgId - ID of the package
+     * @param version - String of the version of the package_version
+     * @return - Value of the package_version if found, -1 otherwise.
+     */
+    private Long getPackageVersionIdFromVersion(Long pkgId, String version) {
+
+        // Find the package version record
+        PackageVersionsRecord pkgVersionRecord = (PackageVersionsRecord) selectedContext.select()
+                .from(PackageVersions.PACKAGE_VERSIONS)
+                .where(PackageVersions.PACKAGE_VERSIONS.PACKAGE_ID.equal(pkgId))
+                .and(PackageVersions.PACKAGE_VERSIONS.VERSION.equal(version))
+                .fetchOne();
+
+        if(pkgVersionRecord != null) {
+            return pkgVersionRecord.getId();
+        }
+
+        return -1L;
```

Comment:
Package version with ID = -1 is also already taken for external callable reference. Same goes for modules and packages.
```sql
fasten_java=> select * from packages where id = -1;
 -1 | external_callables_library | mvn   |              |            | 

fasten_java=> select * from package_versions where id = -1;
 -1 |         -1 | 0.0.1   | OPAL         |            |          | 

fasten_java=> select * from modules where id=-1;
 -1 |                 -1 | global_external_callables |            | 
```

Reply:
OK, done. Null is returned instead of -1L, and checked on it.